File: 6 Kyu Kata/Multiples of 3 or 5/Python/Multiples of 3 or 5.py
#
# * My submitted solution
def solution(n):
    return sum(i for i in range(1, n) if i % 3 == 0 or i % 5 == 0)


# * An alternative solution not submitted, but passed test for submission
def solution2(n):
    res = 0
    for i in range(1, n):
        if i % 3 == 0 or i % 5 == 0:
            res += i
    return res


# * Solution from test cases for reference
def referenceSolution(n):

    if n < 0:
        return 0

    def sum(max, d):
        n = max // d
        s = n * (n + 1) * d // 2
        return s

    n -= 1
    return sum(n, 3) + sum(n, 5) - sum(n, 3 * 5)

File: 6_Kyu_Kata/Multiples_of_3_or_5/TS/MultiplesOf3Or5.ts
//* My submitted solution
// export class Challenge {
function solution(n: number) {
  /* static solution(n: number) { */
  return n > 0 ? [...Array(n).keys()].reduce((a, c, i) => a + (!(i % 3) || !(i % 5) ? c : 0), 0) : 0
}
// }

//* Solution from Kata test case - included for reference
function _solution(n: number) {
  let sum = 0
  for (let i = 0; i < n; i++) {
    if (i % 3 == 0 || i % 5 == 0) {
      sum = sum + i
    }
  }
  return sum
}

exports = module.exports = { solution, _solution }
